File: src/libqgesturerecognizer.cpp
#include <QEvent>
#include <QGesture>
#include <QGestureRecognizer>
#include <QObject>
#include <qgesturerecognizer.h>
#include "libqgesturerecognizer.h"
#include "libqgesturerecognizer.hxx"

QGestureRecognizer* QGestureRecognizer_new() {
    return new VirtualQGestureRecognizer();
}

int QGestureRecognizer_RegisterRecognizer(QGestureRecognizer* recognizer) {
    return static_cast<int>(QGestureRecognizer::registerRecognizer(recognizer));
}

void QGestureRecognizer_UnregisterRecognizer(int typeVal) {
    QGestureRecognizer::unregisterRecognizer(static_cast<Qt::GestureType>(typeVal));
}

void QGestureRecognizer_OperatorAssign(QGestureRecognizer* self, const QGestureRecognizer* param1) {
    self->operator=(*param1);
}

// Derived class handler implementation
QGesture* QGestureRecognizer_Create(QGestureRecognizer* self, QObject* target) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        return vqgesturerecognizer->create(target);
    } else {
        return self->QGestureRecognizer::create(target);
    }
}

// Base class handler implementation
QGesture* QGestureRecognizer_QBaseCreate(QGestureRecognizer* self, QObject* target) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->setQGestureRecognizer_Create_IsBase(true);
        return vqgesturerecognizer->create(target);
    } else {
        return self->QGestureRecognizer::create(target);
    }
}

// Auxiliary method to allow providing re-implementation
void QGestureRecognizer_OnCreate(QGestureRecognizer* self, intptr_t slot) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->setQGestureRecognizer_Create_Callback(reinterpret_cast<VirtualQGestureRecognizer::QGestureRecognizer_Create_Callback>(slot));
    }
}

// Derived class handler implementation
int QGestureRecognizer_Recognize(QGestureRecognizer* self, QGesture* state, QObject* watched, QEvent* event) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        return static_cast<int>(vqgesturerecognizer->recognize(state, watched, event));
    } else {
        return static_cast<int>(((VirtualQGestureRecognizer*)self)->recognize(state, watched, event));
    }
}

// Base class handler implementation
int QGestureRecognizer_QBaseRecognize(QGestureRecognizer* self, QGesture* state, QObject* watched, QEvent* event) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->setQGestureRecognizer_Recognize_IsBase(true);
        return static_cast<int>(vqgesturerecognizer->recognize(state, watched, event));
    } else {
        return static_cast<int>(((VirtualQGestureRecognizer*)self)->recognize(state, watched, event));
    }
}

// Auxiliary method to allow providing re-implementation
void QGestureRecognizer_OnRecognize(QGestureRecognizer* self, intptr_t slot) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->setQGestureRecognizer_Recognize_Callback(reinterpret_cast<VirtualQGestureRecognizer::QGestureRecognizer_Recognize_Callback>(slot));
    }
}

// Derived class handler implementation
void QGestureRecognizer_Reset(QGestureRecognizer* self, QGesture* state) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->reset(state);
    } else {
        self->QGestureRecognizer::reset(state);
    }
}

// Base class handler implementation
void QGestureRecognizer_QBaseReset(QGestureRecognizer* self, QGesture* state) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->setQGestureRecognizer_Reset_IsBase(true);
        vqgesturerecognizer->reset(state);
    } else {
        self->QGestureRecognizer::reset(state);
    }
}

// Auxiliary method to allow providing re-implementation
void QGestureRecognizer_OnReset(QGestureRecognizer* self, intptr_t slot) {
    auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
    if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
        vqgesturerecognizer->setQGestureRecognizer_Reset_Callback(reinterpret_cast<VirtualQGestureRecognizer::QGestureRecognizer_Reset_Callback>(slot));
    }
}

void QGestureRecognizer_Delete(QGestureRecognizer* self) {
    delete self;
}

File: src/libqgesturerecognizer.zig
const C = @import("qt6c");
const qnamespace_enums = @import("libqnamespace.zig").enums;

/// https://doc.qt.io/qt-6/qgesturerecognizer.html
pub const qgesturerecognizer = struct {
    /// New constructs a new QGestureRecognizer object.
    ///
    ///
    pub fn New() ?*C.QGestureRecognizer {
        return C.QGestureRecognizer_new();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qgesturerecognizer.html#create)
    ///
    /// ``` self: ?*C.QGestureRecognizer, target: ?*C.QObject ```
    pub fn Create(self: ?*anyopaque, target: ?*anyopaque) ?*C.QGesture {
        return C.QGestureRecognizer_Create(@ptrCast(self), @ptrCast(target));
    }

    /// Allows for overriding the related default method
    ///
    /// ``` self: ?*C.QGestureRecognizer, slot: fn (?*C.QGestureRecognizer, ?*C.QObject) callconv(.c) ?*C.QGesture ```
    pub fn OnCreate(self: ?*anyopaque, slot: fn (?*anyopaque, ?*anyopaque) callconv(.c) ?*C.QGesture) void {
        C.QGestureRecognizer_OnCreate(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// Base class method implementation
    ///
    /// ``` self: ?*C.QGestureRecognizer, target: ?*C.QObject ```
    pub fn QBaseCreate(self: ?*anyopaque, target: ?*anyopaque) ?*C.QGesture {
        return C.QGestureRecognizer_QBaseCreate(@ptrCast(self), @ptrCast(target));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qgesturerecognizer.html#recognize)
    ///
    /// ``` self: ?*C.QGestureRecognizer, state: ?*C.QGesture, watched: ?*C.QObject, event: ?*C.QEvent ```
    pub fn Recognize(self: ?*anyopaque, state: ?*anyopaque, watched: ?*anyopaque, event: ?*anyopaque) i64 {
        return C.QGestureRecognizer_Recognize(@ptrCast(self), @ptrCast(state), @ptrCast(watched), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    ///
    /// ``` self: ?*C.QGestureRecognizer, slot: fn (?*C.QGestureRecognizer, ?*C.QGesture, ?*C.QObject, ?*C.QEvent) callconv(.c) i64 ```
    pub fn OnRecognize(self: ?*anyopaque, slot: fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) i64) void {
        C.QGestureRecognizer_OnRecognize(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// Base class method implementation
    ///
    /// ``` self: ?*C.QGestureRecognizer, state: ?*C.QGesture, watched: ?*C.QObject, event: ?*C.QEvent ```
    pub fn QBaseRecognize(self: ?*anyopaque, state: ?*anyopaque, watched: ?*anyopaque, event: ?*anyopaque) i64 {
        return C.QGestureRecognizer_QBaseRecognize(@ptrCast(self), @ptrCast(state), @ptrCast(watched), @ptrCast(event));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qgesturerecognizer.html#reset)
    ///
    /// ``` self: ?*C.QGestureRecognizer, state: ?*C.QGesture ```
    pub fn Reset(self: ?*anyopaque, state: ?*anyopaque) void {
        C.QGestureRecognizer_Reset(@ptrCast(self), @ptrCast(state));
    }

    /// Allows for overriding the related default method
    ///
    /// ``` self: ?*C.QGestureRecognizer, slot: fn (?*C.QGestureRecognizer, ?*C.QGesture) callconv(.c) void ```
    pub fn OnReset(self: ?*anyopaque, slot: fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        C.QGestureRecognizer_OnReset(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// Base class method implementation
    ///
    /// ``` self: ?*C.QGestureRecognizer, state: ?*C.QGesture ```
    pub fn QBaseReset(self: ?*anyopaque, state: ?*anyopaque) void {
        C.QGestureRecognizer_QBaseReset(@ptrCast(self), @ptrCast(state));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qgesturerecognizer.html#registerRecognizer)
    ///
    /// ``` recognizer: ?*C.QGestureRecognizer ```
    pub fn RegisterRecognizer(recognizer: ?*anyopaque) i64 {
        return C.QGestureRecognizer_RegisterRecognizer(@ptrCast(recognizer));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qgesturerecognizer.html#unregisterRecognizer)
    ///
    /// ``` typeVal: qnamespace_enums.GestureType ```
    pub fn UnregisterRecognizer(typeVal: i64) void {
        C.QGestureRecognizer_UnregisterRecognizer(@intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qgesturerecognizer.html#operator=)
    ///
    /// ``` self: ?*C.QGestureRecognizer, param1: ?*C.QGestureRecognizer ```
    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        C.QGestureRecognizer_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QGestureRecognizer ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QGestureRecognizer_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qgesturerecognizer.html#types
pub const enums = struct {
    pub const ResultFlag = enum {
        pub const Ignore: i32 = 1;
        pub const MayBeGesture: i32 = 2;
        pub const TriggerGesture: i32 = 4;
        pub const FinishGesture: i32 = 8;
        pub const CancelGesture: i32 = 16;
        pub const ResultState_Mask: i32 = 255;
        pub const ConsumeEventHint: i32 = 256;
        pub const ResultHint_Mask: i32 = 65280;
    };
};
